### DISCUSSION TITLE
uv.pipe_open(stdout, 1) does not work on Windows

### DISCUSSION THREAD

**Discussion by q962** [luvit/luvit#1216] [General]
Use uv_fs_write to solve cross-platform issues.

Set stdout to 1 and then do special processing

```lua
  if type(fs) == 'number' then
    uv.fs_write(fs, s)
  else
    if uv.guess_handle(uv.fileno(fs))=='tty' then
      repeat
        local n, e = uv.try_write(fs, s)
        if n then
          s = s:sub(n+1)
          n = 0
        else
          if e:match('^EAGAIN') then
            n = 0
          else
            assert(n, e)
          end
        end
      until n==#s
    else
      uv.write(fs, s)
    end
  end
```

**Comment by fithisux** [comment]
Can you give a more complete case?

**Reply by q962** [reply]
This code comes from the console_write function in deps/pretty-print.rua.

To allow the program output to be redirected, it needs to be modified.
That is, set stdout to 1 instead of pipe

**Reply by Bilal2453** [reply]
this is still not descriptive enough. What cross-platform issues are you talking about exactly?

pretty-print creates a NEW stdout handle that uses Libuv TTY, for an async interface.  Some platforms do not support TTY handles (terminal handles basically) though, for example in CI, in which case a pipe or a file handle is expected to be set for stdout, This code handles this situation.

**Reply by q962** [reply]
```
stdout = uv.new_pipe(false)
uv.pipe_open(stdout, 1)
```

```
[string "bundle:/deps/pretty-print.lua"]:302: bad argument #1 to 'guess_handle' (number expected, got nil)
stack traceback:
        [C]: in function 'guess_handle'
        [string "bundle:/deps/pretty-print.lua"]:302: in function 'console_write'
        [string "bundle:/deps/pretty-print.lua"]:328: in function 'print'
        ..._\project\demo\clipboard-server\lua_server\local-env.lua:6: in function 'fn'
        [string "bundle:deps/require.lua"]:310: in function 'require'
        [string "bundle:/main.lua"]:128: in function <[string "bundle:/main.lua"]:20>
stack traceback:
        [C]: in function 'error'
        [string "bundle:/deps/utils.lua"]:41: in function 'assertResume'
        [string "bundle:/init.lua"]:52: in function <[string "bundle:/init.lua"]:47>
        [C]: in function 'xpcall'
        [string "bundle:/init.lua"]:47: in function 'fn'
        [string "bundle:deps/require.lua"]:310: in function <[string "bundle:deps/require.lua"]:266>
```


No, this only works on linux.

Whereas uv.fs_write(1, "") is just a basic file write operation, which works on any platform.

**Reply by Bilal2453** [reply]
uv_fs_write cannot write to tty, if stdout is a tty this will fail.  That is why we need to check for the handle being tty or something else. In this case it seems like `uv.fileno` is failing, this is on Windows? Mind opening an issue with the reproduction code?

**Reply by q962** [reply]
The error only occurs when the output is redirected.

sorry I don't quite understand what you mean by your last sentence.

**Reply by q962** [reply]
Oh, maybe you want me to open an issue instead of continuing to use discussions?

**Reply by Bilal2453** [reply]
Yes, with steps of how to get this error so I can figure out the issue

**Reply by q962** [reply]
ok, but I need to go to bed. I'll open an issue tomorrow.

But before that, you can take a look at
https://github.com/libuv/libuv/discussions/4587#discussioncomment-10992979
